File: Gombang/[LeetCode75] - [Array_Sort]/443_String_Compression.cpp
class Solution {
public:
    int compress(vector<char>& chars) {

        const int DEFAULT_COUNT = 1;
        vector<char> result;
        int count = DEFAULT_COUNT;
        int currChar = chars[0];

        for (int i = 1; i < chars.size(); i++)
        {
            if (currChar == chars[i])
            {
                count++;
            }
            else
            {
                result.push_back(currChar);
                currChar = chars[i]; // currChar 변경.

                // count가 1보다 클 경우에만 vector에 푸쉬하기 위한 예외처리.
                if (count <= 1)
                    continue;

                string countToString = to_string(count);
                for (int j = 0; j < countToString.size(); j++)
                {
                    result.push_back(countToString[j]);
                }

                count = DEFAULT_COUNT;
            }
        }

        // 마지막 요소 삽입.
        result.push_back(currChar);
        if (count > 1)
        {
            string countToString = to_string(count);
            for (int i = 0; i < countToString.size(); i++)
            {
                result.push_back(countToString[i]);
            }
        }

        chars = result;
        return chars.size();
    }
};

File: Seori/LeetCode 75/Array_String/443. String Compression.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        count = 1
        character = chars[0]
        answer = 0
        answer_chars = ''
        length = len(chars)
        for i in range(1, length + 1):
            # [1] 앞뒤 문자가 같은 경우 count만 증가시킨다.
            if i < length and character == chars[i]:
                count += 1
            # [2] 앞뒤 문자가 다른 경우 count와 character를 answer에 반영하고 초기화한다.
            else:
                answer_chars += character
                answer += 1
                if count > 1:
                    answer_chars += str(count)
                    while count:
                        count //= 10
                        answer += 1
                if i < length:
                    character = chars[i]
                count = 1
        
        # [3] answer에 저장된 값으로 chars를 갱신한다.
        for i in range(answer):
            chars[i] = answer_chars[i]
        return answer